extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/OAuth2Template.java

```diff
@@ -0,0 +1,115 @@
+        SSLContext sslContext = config.caCertFile.isPresent() ? createSLLContext(config) : SSLContext.getDefault();
+
+        TokenValidator validator = OAuth2IntrospectValidator.builder()
+                .clientId(config.clientId)
+                .clientSecret(config.clientSecret)
+                .tokenIntrospectionUrl(URI.create(config.introspectionUrl).toURL())
+                .useSslContext(sslContext)
+                .build();
+
+        TokenSecurityRealm tokenRealm = TokenSecurityRealm.builder()
+                .validator(validator)
+                .claimToPrincipal(claims -> new ElytronOAuth2CallerPrincipal(claims))
+                .build();
+
+        return new RuntimeValue<>(tokenRealm);
+    }
+
+    private SSLContext createSLLContext(OAuth2Config config)
+            throws IOException, CertificateException, NoSuchAlgorithmException, KeyStoreException, KeyManagementException {
+        InputStream is = new FileInputStream(config.caCertFile.get());
+
+        CertificateFactory cf = CertificateFactory.getInstance("X.509");
+        X509Certificate caCert = (X509Certificate) cf.generateCertificate(is);
+
+        TrustManagerFactory tmf = TrustManagerFactory
+                .getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
+        ks.load(null); // You don't need the KeyStore instance to come from a file.
+        ks.setCertificateEntry("caCert", caCert);
+
+        tmf.init(ks);
+
+        SSLContext sslContext = SSLContext.getInstance("TLS");
+        sslContext.init(null, tmf.getTrustManagers(), null);
+
+        return sslContext;
+    }
+
+    /**
+     * Create an OAuth2IdentityManager for the given SecurityDomain
+     *
+     * @param domain - configured SecurityDomain
+     * @return runtime value for OAuth2IdentityManager
+     */
+    public IdentityManager createIdentityManager(RuntimeValue<SecurityDomain> domain) {
+        return new OAuth2IdentityManager(domain.getValue());
+    }
+
+    /**
+     * Called to create a {@linkplain ServletExtension} to associate the {@linkplain ElytronIdentityManager} with the
+     * deployment.
+     *
+     * @param domain - the SecurityDomain to use for auth decisions
+     * @param identityManager - the IdentityManager for auth decisions
+     * @return - the ServletExtension instance to register
+     */
+    public ServletExtension configureUndertowIdentityManager(RuntimeValue<SecurityDomain> domain,
+            IdentityManager identityManager) {
+        return new ServletExtension() {
+            @Override
+            public void handleDeployment(DeploymentInfo deploymentInfo, ServletContext servletContext) {
+                deploymentInfo.setIdentityManager(identityManager);
+
+                //TODO maybe we need to include it only when oauth2 is enabled
+                deploymentInfo.addAuthenticationMechanism("BEARER_TOKEN", new OAuth2AuthMechanismFactory());
```

Comment:
Indeed, it probably makes sense to avoid making the authMechanism type/name configurable for now

Reply:
removed